### ISSUE TITLE
IncrementalPCA

### ISSUE THREAD

**Issue by fujiisoup** [dask/dask-ml#617]
Hi all,

Although dask-ml already has PCA and wrappers.Incremented, but there is no IncrementalPCA.
Is this in the scope of development?

With dask-ml's PCA, we can handle large data, but if the data is too large, some approximated algorithms may be necessary.
With wrappers.Incremental, we could use scikit-learn's IncrementalPCA, but it internally uses scipy.linalg.svd, which won't run with large feature size of data.

For me it would be nicer if dask-ml has dask-compatible IncrementalPCA.

I implemented it in my folk, most part of which is stolen from scikit-learn but uses dask.array.linalg.svd_compressed internally.
If it is in the scope, I will clean it a bit and send a PR.

Thank you!

cc:
@demaheim
@yasahi-hpc

**Comment by TomAugspurger** [comment]
>  With wrappers.Incremental, we could use scikit-learn's IncrementalPCA, but it internally uses scipy.linalg.svd, which won't run with large feature size of data.

I think something will work, if you're OK training serially in batches.

>  If it is in the scope, I will clean it a bit and send a PR.

Yes, this would be in scope and most welcome! It'd avoid some of the drawbacks of using `wrappers.Incremental(sklearn.decomposition.IncrementalPCA)`.

**Comment by stsievert** [comment]
>  It'd avoid some of the drawbacks of using `wrappers.Incremental(sklearn.decomposition.IncrementalPCA)`

What are those drawbacks?

I wasn't aware Scikit-learn had an `IncrementalPCA`! That makes this go from "relevant" to "very relevant".

**Comment by TomAugspurger** [comment]
I haven't looked, but I suspect that `wrappers.Incremental(sklearn.decomposition.IncrementalPCA)` will be fit serially. If an dask-ml IncrementalPCA used `dask.array.linalg.svd_compressed`, that would be done in parallel.

I'm also not sure if it matters, but I suspect fitting the SVD on the whole dataset and then updating the parameters is preferable to fitting on a batch and updating the parameters many times.

**Comment by fujiisoup** [comment]
Scikit-learn's IncrementalPCA works well if data has a large sample size but smaller feature size.
It does SVD to a matrix shaped (batch_size x feature_size) and updates parameters.
But if the data has a large feature size, SVD of (batch_size x feature_size) won't work or at least takes a lot of time.
This part I want to parallelize.
This will allow us to user larger batch_size.

Another advantage of having our own IncrementalPCA rather than using `dask_ml.Incremental` migh be that we can store an intermediate learning state and make a hot start if new data arrives.
(Probably can it be done even with our Incremental wrappers?)
